File: 0020_valid_parentheses/solution01.cpp
#include <unordered_map>
#include <unordered_set>
#include <vector>

// runtime: 100.00%, memory: 15.02% 
class Solution {
public:
    bool isValid(string s) {
        int l = s.length();
        if (l < 2 || l % 2 == 1) {
            return false;
        }

        unordered_set<char> parens = {')', '}', ']'};
        unordered_map<char, char> pairs = { 
            {')', '('}, {'}', '{'}, {']', '['}};
        vector<char> stack;

        for (char ch : s) {
            if (stack.size() > 0 && parens.contains(ch)) {
                if (stack.back() == pairs[ch]) {
                    stack.pop_back();
                } else {
                    stack.push_back(ch);
                }
            } else {
                stack.push_back(ch);
            }
        }

        return stack.size() == 0;
    }
};

File: 0020_valid_parentheses/solution01.py
from collections import deque 

# runtime: 100.00%, memory: 57.01%
class Solution:
    def isValid(self, s: str) -> bool: 
        parens = {
            ")": "(",
            "}": "{",
            "]": "[",
        }
        stack = deque()

        for ch in s:
            if stack and ch in parens:
                if stack[-1] == parens[ch]:
                    stack.pop()
                else:
                    stack.append(ch)
            else: 
                stack.append(ch)

        return len(stack) == 0 

        